dns_sd_windows.c

```diff
@@ -0,0 +1,262 @@
+					memcmp(saddr->sin6_addr.s6_addr, localhost, 16) &&
+					memcmp(saddr->sin6_addr.s6_addr, localhost_mapped, 16)) {
+					if (num_sockets < max_sockets) {
+						int sock = mdns_socket_open_ipv6(saddr);
+						if (sock >= 0) {
+							sockets[num_sockets++] = sock;
+						}
+					}
+				}
+			}
+		}
+	}
+
+	free(adapter_address);
+#endif
+
+	for (int isock = 0; isock < num_sockets; ++isock) {
+#ifdef _WIN32
+		unsigned long param = 1;
+		ioctlsocket(sockets[isock], FIONBIO, &param);
+#else
+		const int flags = fcntl(sockets[isock], F_GETFL, 0);
+		fcntl(sockets[isock], F_SETFL, flags | O_NONBLOCK);
+#endif
+	}
+
+	return num_sockets;
+}
+
+
+static int
+query_callback(int sock, const struct sockaddr* from, size_t addrlen,
+	mdns_entry_type_t entry, uint16_t transaction_id,
+	uint16_t rtype, uint16_t rclass, uint32_t ttl,
+	const void* data, size_t size, size_t offset, size_t length,
+	void* user_data) {
+
+
+	char addrbuffer[64];
+	char servicebuffer[64];
+	char namebuffer[256];
+
+	struct dns_sd_discovery_data* dd = (struct dns_sd_discovery_data*)user_data;
+	if (dd == NULL) {
+		IIO_ERROR("DNS SD: Missing info structure. Stop browsing.\n");
+		goto quit;
+	}
+
+	if (rtype != MDNS_RECORDTYPE_SRV)
+		goto quit;
+
+	getnameinfo((const struct sockaddr*)from, (socklen_t)addrlen,
+		addrbuffer, NI_MAXHOST, servicebuffer, NI_MAXSERV,
+		NI_NUMERICSERV | NI_NUMERICHOST);
+
+	mdns_record_srv_t srv = mdns_record_parse_srv(data, size, offset, length,
+		namebuffer, sizeof(namebuffer));
+	IIO_DEBUG("%s : SRV %.*s priority %d weight %d port %d\n",
+		addrbuffer,
+		MDNS_STRING_FORMAT(srv.name), srv.priority, srv.weight, srv.port);
+
+	// Go to the last element in the list
+	while (dd->next != NULL)
+		dd = dd->next;
+
+	if (srv.name.length > 1)
+	{
+		dd->hostname = malloc(srv.name.length);
+		strncpy(dd->hostname, srv.name.str, srv.name.length);
+		dd->hostname[srv.name.length - 1] = 0;
+	}
+	strcpy(dd->addr_str, addrbuffer);
+	dd->port = srv.port;
+
+	IIO_DEBUG("DNS SD: added %s (%s:%d)\n", dd->hostname, dd->addr_str, dd->port);
+	// A list entry was filled, prepare new item on the list.
+	if (new_discovery_data(&dd->next)) {
+		IIO_ERROR("DNS SD mDNS Resolver : memory failure\n");
+	}
+
+quit:
+	return 0;
+}
+
+int dnssd_find_hosts(struct dns_sd_discovery_data** ddata)
+{
+#ifdef _WIN32
+	const char* hostname = "dummy-host";
+	WORD versionWanted = MAKEWORD(1, 1);
+	WSADATA wsaData;
+	if (WSAStartup(versionWanted, &wsaData)) {
+		printf("Failed to initialize WinSock\n");
+		return -1;
+	}
+
+	char hostname_buffer[128];
+	DWORD hostname_size = (DWORD)sizeof(hostname_buffer);
+	if (GetComputerNameA(hostname_buffer, &hostname_size))
+		hostname = hostname_buffer;
+#endif
+
+	struct dns_sd_discovery_data* d;
+
+	IIO_DEBUG("DNS SD: Start service discovery.\n");
+
+	if (new_discovery_data(&d) < 0) {
+		return -ENOMEM;
+	}
+	*ddata = d;
+
+	size_t capacity = 2048;
+	void* buffer = malloc(capacity);
+	const char service[] = "_iio._tcp.local.";
```

Comment:
In the other implementation we use:
```
dns_sd_avahi.c:			"_iio._tcp", NULL, 0, __avahi_browser_cb, d);
dns_sd_bonjour.c:	CFStringRef type = CFSTR("_iio._tcp.");
./iiod/iiod.c:			0, avahi.name, "_iio._tcp", NULL, NULL, IIOD_PORT, NULL);
```

no fully qualified domain. I'm not sure which is 100% correct - but we should be doing the same for all.

Reply:
I don't know what the correct version is, but not appending .local here makes it not detect anything.

Reply:
Ok - I guess that is the correct version :) Maybe the .local is appended in side the other libraries.